### PR TITLE
Change the device library to use CUDA runtime API

### PR DESCRIPTION
**Pull request by alexrlongne**
### Background

* Our CUDA device helper classes used a the older CUDA driver API, which is not recommended for new projects. The new runtime API is much simpler and doesn't require explicit management of modules or contexts. It also provides some very nice helper functions to diagnose runtime errors and query the device. This PR changes the `GPU_Device` helper class to use these newer functions. It also demonstrates how to compile a library of CUDA functions that can be called from either the host or device. To linking to a library of CUDA device functions the target library must be static. This required a slight change to the `add_component_library` CMake macro.

### Purpose of Pull Request

* [Completes Redmine Issue #1429](https://rtt.lanl.gov/redmine/issues/1429)

### Description of changes

* Add a simple CUDA compilation step during configure to determine the appropriate architecture flags
* Allow add_component_library to accept a library type different than the DRACO_LIBRARY_TYPE variable
* Change the GPU_Device class to use the newer CUDA Runtime API
* Remove the GPU_Module class (explicit module and contexts are not necessary in the Runtime API)
* Remove "EXTERN C" decorations (these are also no longer necessary in the Runtime API since functions do not need to be loaded explicitly)
* Add a test that shows an example of calling code from both the host and device
* Consolidate calls to simple GPU test kernels into a "basic_kernels" header and implementation file
* Compile auxiliary GPU test code into a static library callable by the individual gpu tests

### Status

* Reference:
  * [Pre-Merge Code Review](https://github.com/lanl/Draco/wiki/Style-Guide)
  * [CDash Status](https://rtt.lanl.gov/cdash/index.php?project=Draco&filtercount=1&showfilters=1&field1=buildname&compare1=63&value1=-pr)
* Checks
  * [x] Travis CI checks pass
  * [x] Code coverage does not decrease
  * [ ] [Valgrind test passes](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [ ] [Toss3 checks pass](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [ ] [Trinitite checks pass](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [ ] Code reviewed/approved, sufficient DbC checks, testing, documentation

### CODE REVIEW

**Comment by alexrlongne** [comment]
@KineticTheory @jhchang-lanl I'm open to any idea on how to name GPU tests, libraries, files etc. Just let me know and I'll change it.

File: src/device/GPU_Device.cc [hunk 1]
```diff
@@ -9,6 +9,7 @@
 //---------------------------------------------------------------------------//
 
 #include "GPU_Device.hh"
+#include <iostream>
```
Comment (KineticTheory) [inline src/device/GPU_Device.cc]: Is this required?  I thought we prevented this class from writing directly to stdin/stderr?

File: src/device/GPU_Device.cc [hunk 2]
```diff
@@ -26,143 +27,75 @@ namespace rtt_device {
...
-  // Free reserved contexts:
-  for (int device = 0; device < deviceCount; device++) {
-    // Only initialize if compute capability >= 2.0
-    if (computeCapability[device][0] >= 2) {
-      cudaError_enum err = cuCtxDestroy(context[device]);
-      checkForCudaError(err);
-    }
-  }
-}
+GPU_Device::~GPU_Device() {}
```
Comment (KineticTheory) [inline src/device/GPU_Device.cc]: So cuda no longer requires initialization and destruction to use a device?  That is a real improvement!

File: src/device/GPU_Device.cc [hunk 3]
```diff
@@ -218,152 +151,141 @@ void GPU_Device::checkForCudaError(cudaError_enum const errorCode) {
...
   return;
 }
 
-void GPU_Device::MemFree(CUdeviceptr ptr) {
-  cudaError_enum err = cuMemFree(ptr);
+void GPU_Device::MemFree(void *ptr) {
+  cudaError_t err = cudaFree(ptr);
   checkForCudaError(err);
   return;
 }
```
Comment (KineticTheory) [inline src/device/GPU_Device.cc]: Great work reconstructing this file!  I don't like the `void*` stuff, but if that is what is required, then we'll live with it.

File: src/device/GPU_Device.cmake [hunk 4]
```diff
@@ -38,11 +39,14 @@ add_component_library(
    TARGET       Lib_device
    TARGET_DEPS  Lib_dsxx
    LIBRARY_NAME device
+   LIBRARY_TYPE STATIC
```
Comment (KineticTheory) [inline src/device/GPU_Device.cmake]: Is `STATIC` required?  I thought you determined that this wasn't needed?

File: src/device/GPU_Device.cmake [hunk 5]
```diff
@@ -15,6 +15,7 @@
 
 set( TEST_KERNEL_BINDIR ${PROJECT_BINARY_DIR}/test CACHE PATH
   "GPU kernel binary install location" )
+set( CUDA_DEVICE ON)
```
Comment (KineticTheory) [inline src/device/GPU_Device.cmake]: What does this variable do?  Who needs access to it?

Ok - I see that it is used by `config.h`.  Maybe add a comment here to describe how it is used.

File: src/device/GPU_Device.hh [hunk 6]
```diff
@@ -163,16 +152,6 @@ private:
...
-
-  //! Device handles (one per device)
-  std::vector<CUdevice> device_handle;
-  /*! Device context (one per handle)
-   *
-   * Current implementation only allows 1 context per GPU.  However, the CUDA
-   * Driver API provides for the concept of pushing and poping various contexts
-   * on the GPU.
-   */
-  std::vector<CUcontext> context;
```
Comment (KineticTheory) [inline src/device/GPU_Device.hh]: I love seeing this code removed!

File: src/device/device_cuda.h [hunk 7]
```diff
@@ -27,6 +27,7 @@
 #endif
 
 #include <cuda.h>
+#include <cuda_runtime.h>
```
Comment (KineticTheory) [inline src/device/device_cuda.h]: Is it okay to have both of these cuda headers included?  Are both needed?

File: src/device/test/Dual_Call.cu [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+ * \brief  Show how code can be called from GPU and host
+ * \note   Copyright (C) 2019 Triad National Security, LLC.
+ *         All rights reserved.
+ */
+//---------------------------------------------------------------------------//
+
+#include "Dual_Call.hh"
+
+namespace rtt_device_test {
+
```
Comment (KineticTheory) [inline src/device/test/Dual_Call.cu]: Please add doxygen-style comment blocks for each function.

File: src/device/test/GPU_Device.cmake [hunk 9]
```diff
@@ -11,37 +11,42 @@
...
-# target_link_libraries( Ut_gpu_hello_rt_api_exe Lib_device_test Lib_dsxx )
-#target_link_libraries( Ut_gpu_hello_rt_api_exe Lib_device )
+add_component_library(
+   TARGET       Lib_device_test
+   TARGET_DEPS  Lib_dsxx Lib_device
+   LIBRARY_NAME device_test
+   LIBRARY_TYPE STATIC
+   SOURCES      "${cuda_sources}"
+   HEADERS      "${cuda_headers}" )
+set_property(TARGET Lib_device_test PROPERTY CUDA_SEPARABLE_COMPILATION ON)
```
Comment (KineticTheory) [inline src/device/test/GPU_Device.cmake]: Awesome sauce!

* Is STATIC required? If so, can we make this automatic for `*.cu` files?
* Do you think we can move the `CUDA_SEPARABLE_COMPILATION` into `add_component_library`, maybe enable if any source file matches `*.cu`?

File: src/device/test/basic_kernels.hh [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+
+#include "device/config.h"
+#include <algorithm>
+#include <numeric>
+#include <stdio.h>
+#include <vector>
+
+namespace rtt_device_test {
+
+__global__ void vector_add(double const *A_dev, double const *B_dev,
```
Comment (KineticTheory) [inline src/device/test/basic_kernels.hh]: Please add doxygen-style comments.

File: src/device/test/gpu_device_info.cu [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+// Main
+//---------------------------------------------------------------------------//
+
+int main(int argc, char *argv[]) {
+  using namespace std;
+
+  rtt_dsxx::ScalarUnitTest ut(argc, argv, rtt_dsxx::release);
+  try {
+    query_device(ut);
+  } catch (exception &err) {
```
Comment (KineticTheory) [inline src/device/test/gpu_device_info.cu]: Replace with UT_EPILOG?

File: src/device/test/gpu_device_info.cu [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+
+  // Print the message to stdout
+  cout << out.str();
+
+  // Parse the output
+  bool verbose(false);
+  std::map<std::string, unsigned> wordCount =
+      rtt_dsxx::get_word_count(out, verbose);
+
+  if (wordCount[string("Device")] == numDev)
```
Comment (KineticTheory) [inline src/device/test/gpu_device_info.cu]: Use FAIL_IF_NOT?

File: config/vendor_libraries.cmake [hunk 13]
```diff
@@ -299,6 +299,32 @@ macro( setupCudaEnv )
...
+    if(${CUDA_RETURN_CODE} EQUAL 0)
+      set(CUDA_SUCCESS "TRUE")
+    else()
+      set(CUDA_SUCCESS "FALSE")
+    endif()
+
+  if (${CUDA_SUCCESS})
+    message(STATUS "CUDA Architecture: ${ARCH}")
+    set(CMAKE_CUDA_FLAGS "${ARCH} -G -O0")
+    # ARL: These variables are set by the find_package(CUDA) call, which we
```
[the remark this answers is not in the document]
Comment (alexrlongne) [reply config/vendor_libraries.cmake]: I can do that. That's right, I see that find_cuda is deprecated in the newest CMake.

**Comment by KineticTheory** [comment]
@KineticTheory @clevelam Is email from gitlab working

File: config/vendor_libraries.cmake [hunk 14]
```diff
@@ -299,6 +299,32 @@ macro( setupCudaEnv )
...
+
+    if(${CUDA_RETURN_CODE} EQUAL 0)
+      set(CUDA_SUCCESS "TRUE")
+    else()
+      set(CUDA_SUCCESS "FALSE")
+    endif()
+
+  if (${CUDA_SUCCESS})
+    message(STATUS "CUDA Architecture: ${ARCH}")
+    set(CMAKE_CUDA_FLAGS "${ARCH} -G -O0")
```
[the remark this answers is not in the document]
Comment (alexrlongne) [reply config/vendor_libraries.cmake]: * Yes, I'll cache these flags
* I've made flags for each build type. There are other numerical optimization flags (e.g. --fast-math) we can discuss later and this page (https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/#options-for-altering-compiler-linker-behavior) has everything we might need for Debug and RelWithDebug builds.

File: config/vendor_libraries.cmake [hunk 15]
```diff
@@ -299,6 +299,32 @@ macro( setupCudaEnv )
...
       "If CUDA is available, this variable is 'CUDA'")
 
+      set(OUTPUTFILE ${CMAKE_CURRENT_SOURCE_DIR}/config/cuda_script) # No suffix required
+      set(CUDAFILE ${CMAKE_CURRENT_SOURCE_DIR}/config/query_gpu.cu)
+      execute_process(COMMAND nvcc -lcuda ${CUDAFILE} -o ${OUTPUTFILE})
+      execute_process(COMMAND ${OUTPUTFILE}
+                      RESULT_VARIABLE CUDA_RETURN_CODE
+                  OUTPUT_VARIABLE ARCH)
+
+    if(${CUDA_RETURN_CODE} EQUAL 0)
```
[the remark this answers is not in the document]
Comment (alexrlongne) [reply config/vendor_libraries.cmake]: Done, thanks!

File: config/vendor_libraries.cmake (hunk 14 above)
[the remark this answers is not in the document]
Comment (alexrlongne) [reply config/vendor_libraries.cmake]: Flags changed.

File: config/vendor_libraries.cmake [hunk 16]
```diff
@@ -299,6 +299,32 @@ macro( setupCudaEnv )
     set( CUDA_DBS_STRING "CUDA" CACHE BOOL
       "If CUDA is available, this variable is 'CUDA'")
 
+      set(OUTPUTFILE ${CMAKE_CURRENT_SOURCE_DIR}/config/cuda_script) # No suffix required
+      set(CUDAFILE ${CMAKE_CURRENT_SOURCE_DIR}/config/query_gpu.cu)
+      execute_process(COMMAND nvcc -lcuda ${CUDAFILE} -o ${OUTPUTFILE})
+      execute_process(COMMAND ${OUTPUTFILE}
+                      RESULT_VARIABLE CUDA_RETURN_CODE
+                  OUTPUT_VARIABLE ARCH)
```
[the remark this answers is not in the document]
Comment (alexrlongne) [reply config/vendor_libraries.cmake]: Fixed. Let me know if it still looks weird (or if you have a specific style for CMake files).

File: config/vendor_libraries.cmake [hunk 17]
```diff
@@ -299,6 +299,32 @@ macro( setupCudaEnv )
     set( CUDA_DBS_STRING "CUDA" CACHE BOOL
       "If CUDA is available, this variable is 'CUDA'")
 
+      set(OUTPUTFILE ${CMAKE_CURRENT_SOURCE_DIR}/config/cuda_script) # No suffix required
```
[the remark this answers is not in the document]
Comment (alexrlongne) [reply config/vendor_libraries.cmake]: Added an include guard using a variable that's defined in the if_cuda block.

File: src/device/GPU_Device.cc (hunk 3 above)
Comment (alexrlongne) [reply src/device/GPU_Device.cc]: Yeah, the `CUdeviceptr` is part of the driver API. It's defined as:

"CUDA device pointer CUdeviceptr is defined as an unsigned integer type whose size matches the size of a pointer on the target platform."

I'm not sure how different that was from a raw pointer?

File: src/device/GPU_Device.cmake (hunk 5 above)
Comment (alexrlongne) [reply src/device/GPU_Device.cmake]: I can add a comment. Should I cache this as variable as well?

File: src/device/test/GPU_Device.cmake (hunk 9 above)
Comment (alexrlongne) [reply src/device/test/GPU_Device.cmake]: * Yes, static is required as far as I can tell. I found that in this presentation (http://on-demand.gputechconf.com/gtc-express/2012/presentations/gpu-object-linking.pdf). Yes, automatically doing this for .cu files seems like a great idea!
* I think we can do this through the same mechanism as the automatic static build.
 
Do you want me to add those features as part of this PR?

File: src/device/test/basic_kernels.hh (hunk 10 above)
Comment (alexrlongne) [reply src/device/test/basic_kernels.hh]: Done.

File: src/device/test/gpu_device_info.cu (hunk 11 above)
Comment (alexrlongne) [reply src/device/test/gpu_device_info.cu]: Yes, that makes it much cleaner.

File: src/device/test/gpu_device_info.cu (hunk 12 above)
Comment (alexrlongne) [reply src/device/test/gpu_device_info.cu]: Done. Thanks!

**Comment by KineticTheory** [comment]
Some weirdness...

```
Preprocessing /home/travis/Dr/home/travis/Draco/src/device/GPU_Device.hh:55: error: included file device/test/gpu_hello_driver_api.cc is not found. Check your EXAMPLE_PATH (warning treated as error, aborting now)
```

I'll help you fix this one.

File: src/device/GPU_Device.hh [hunk 18]
```diff
@@ -106,17 +106,6 @@ public:
    *  textureAlign bytes do not need an offset applied to texture fetches */
   int textureAlign(int devId = 0) const { return m_texturealign[devId]; }
```
Comment (KineticTheory) [inline src/device/GPU_Device.hh]: At the top of this file, replace this doxygen example reference with one of the new files.

```
 * \example device/test/gpu_hello_driver_api.cc
 * Test of GPU_Device for CUDA driver environment.
```

This should fix the autodoc build failure.

File: src/device/test/Dual_Call.cu (hunk 8 above)
Comment (alexrlongne) [reply src/device/test/Dual_Call.cu]: Done. Thanks!

File: src/device/GPU_Device.hh (hunk 18 above)
Comment (alexrlongne) [reply src/device/GPU_Device.hh]: I've removed this reference to a test file for the old driver API.

File: src/device/test/GPU_Device.cmake (hunk 9 above)
Comment (alexrlongne) [reply src/device/test/GPU_Device.cmake]: I'll add a redmine ticket to track this.

**Comment by alexrlongne** [comment]
@KineticTheory I've tried to address all of your comments and the Travis CI is passing. This should be ready for merge once the CDash is clean.

**Pull request merged**